### PR TITLE
RD-4469 Add support for setting expiration date when creating a new token

### PR DESCRIPTION
**Pull request by qooban**
## Description
According to RD-4469 this PR adds support for setting expiration date when creating a new token. Only absolute time setting is provided (there's another ticket for relative time support - RD-5335).

This PR builds on top of https://github.com/cloudify-cosmo/cloudify-stage/pull/2057, so it's required to be merged first.

In addition to main changes I have extended `useErrors` hook to provide a common way of setting context errors in the forms.

## Screenshots / Videos

https://user-images.githubusercontent.com/5202105/178677287-454a62b4-1343-4558-a9c5-eea16b2c0019.mp4

## Checklist
- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I verified that all tests and checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests
1. Updated Tokens widget spec.
2. [Build Status](https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/3111/) - only Tokens widget spec - PASSED

## Documentation
Created https://github.com/cloudify-cosmo/docs.getcloudify.org/pull/2245 . Will sync this PR once it gets approval.

### CODE REVIEW

File: app/translations/en.json [hunk 1]
```diff
@@ -1659,7 +1660,11 @@
                     "content": "New token has been created: {{token}}",
                     "warning": "You won't be able to show or copy the token value after closing this modal"
                 },
-                "error": "Couldn't create a Token, please try to submit the Token data again."
+                "errors": {
+                    "createError": "Couldn't create a Token, please try to submit the Token data again.",
+                    "expirationDateInvalidFormat": "Please provide a date in 'YYYY-MM-DD HH:mm' format",
+                    "expirationDateBeforeCurrentDate": "Please provide a date later than the current date"
```
Comment (NorbertAmartus) [inline app/translations/en.json:1666]:
```suggestion
                    "expirationDateBeforeCurrentDate": "Please provide a date later than the present date"
```

File: app/utils/hooks/useErrors.ts [hunk 2]
```diff
@@ -1,14 +1,19 @@
 import { useState } from 'react';
+import type { StrictLabelProps } from 'semantic-ui-react';
 
 /**
- * Returns a stateful `errors` object and functions to manipulate it: `setErrors`, `clearErrors` and `setMessageAsError`,
- * which extracts message from given `Error` instance and sets it as actual error
+ * Returns a stateful `errors` object and functions to manipulate it: `setErrors`, `clearErrors` and helper functions:
+ * `setMessageAsError`, which extracts message from given `Error` instance and sets it as actual error,
+ * `getContextError`, which for a given field name provides a value to be passed to `error` prop in `Form.Field`
+ * component
```
Comment (NorbertAmartus) [inline app/utils/hooks/useErrors.ts:8]:
```suggestion
 * `getContextError`, which for a given field name provides a value to be passed to `error` prop in `Form.Field` component
```

File: widgets/tokens/src/TokensTableHeader/CreateTokenModal.tsx [hunk 3]
```diff
@@ -1,37 +1,64 @@
...
 
 const t = getT(`${translationPath}.createModal`);
+const expirationDateFormat = 'YYYY-MM-DD HH:mm';
+
+function getTokensPostRequestBody(description: string, expirationDate: string): TokensPostRequestBody {
+    const body: TokensPostRequestBody = {};
+    if (description) body.description = description;
+    if (expirationDate) body.expiration_date = moment.utc(moment(expirationDate)).format(expirationDateFormat);
+    return body;
+}
```
Comment (NorbertAmartus) [inline widgets/tokens/src/TokensTableHeader/CreateTokenModal.tsx:22]: We could extract 
> moment.utc(moment(expirationDate)).format(expirationDateFormat);

As a separate function, to make things more robust 🎉 

```suggestion
function formatRequestExpirationDate(expirationDate: string) {
    return moment.utc(moment(expirationDate)).format(expirationDateFormat);
}

function getTokensPostRequestBody(description: string, expirationDate: string): TokensPostRequestBody {
    const body: TokensPostRequestBody = {};
    if (description) body.description = description;
    if (expirationDate) body.expiration_date = formatRequestExpirationDate(expirationDate);
    return body;
}
```

File: widgets/tokens/src/TokensTableHeader/CreateTokenModal.tsx [hunk 4]
```diff
@@ -1,37 +1,64 @@
...
+            const expirationDateHasInvalidFormat =
+                !expirationDateMoment.isValid() ||
+                !isEqual(expirationDateMoment.format(expirationDateFormat), expirationDate);
+
+            if (expirationDateHasInvalidFormat) {
+                setErrors({ expirationDate: t('errors.expirationDateInvalidFormat') });
+                return;
+            }
+
+            if (expirationDateMoment.isBefore(moment())) {
```
Comment (NorbertAmartus) [inline widgets/tokens/src/TokensTableHeader/CreateTokenModal.tsx:52]: We can extract additional condition as a variable, to make the if'ology more readable 🎉 

```suggestion
            const expirationDateIsPast = expirationDateMoment.isBefore(moment());
            
            if (expirationDateHasInvalidFormat) {
                setErrors({ expirationDate: t('errors.expirationDateInvalidFormat') });
                return;
            }

            if (expirationDateIsPast) {
```

File: app/translations/en.json (hunk 1 above)
Comment (qooban) [reply app/translations/en.json:1666]: Seems like both ways are acceptable. Can you elaborate more on that?

File: widgets/tokens/src/TokensTableHeader/CreateTokenModal.tsx (hunk 4 above)
Comment (qooban) [reply widgets/tokens/src/TokensTableHeader/CreateTokenModal.tsx:52]: Didn't do that intentionally as I thought that `expirationDateMoment.isBefore(moment())` is clear enough. But OK, I'll change it.

File: widgets/tokens/src/TokensTableHeader/CreateTokenModal.tsx (hunk 3 above)
Comment (qooban) [reply widgets/tokens/src/TokensTableHeader/CreateTokenModal.tsx:22]: OK, will change.

File: app/utils/hooks/useErrors.ts (hunk 2 above)
Comment (qooban) [reply app/utils/hooks/useErrors.ts]: Fixed.

File: app/translations/en.json (hunk 1 above)
Comment (NorbertAmartus) [reply app/translations/en.json]: From what I know (but I may be mistaken 😅):
- "current" is used in sense of "recent"
- "present" is used in sense of "at the moment"

The change is not a requirement from my side (as I'm not an English guru 😁)
So as other changes have been made, I'm approving this PR 🎉

**Review by NorbertAmartus — APPROVED** [review]

File: app/translations/en.json (hunk 1 above)
Comment (qooban) [reply app/translations/en.json]: OK, I'll leave it as is then.

See:
1. https://context.reverso.net/t%C5%82umaczenie/angielski-polski/current+date
2. https://context.reverso.net/t%C5%82umaczenie/angielski-polski/present+date

**Pull request merged**